In [1]:
import pandas as pd

In [2]:
# 모든 데이터 로드
df_bibim1 = pd.read_csv("../../selenium/yohan/data/bibim/local.bibim1.csv")
df_bibim2 = pd.read_csv("../../selenium/yohan/data/bibim/local.bibim2.csv")
df_bibim3 = pd.read_csv("../../selenium/yohan/data/bibim/local.bibim3.csv")
df_bibim4 = pd.read_csv("../../selenium/yohan/data/bibim/local.bibim4.csv")

df_oil1 = pd.read_csv("../../selenium/yohan/data/chamoil/local.oil1.csv")
df_oil2 = pd.read_csv("../../selenium/yohan/data/chamoil/local.oil2.csv")
df_oil3 = pd.read_csv("../../selenium/yohan/data/chamoil/local.oil3.csv")
df_oil4 = pd.read_csv("../../selenium/yohan/data/chamoil/local.oil4.csv")
df_oil5 = pd.read_csv("../../selenium/yohan/data/chamoil/local.oil5.csv")

df_ham1 = pd.read_csv("../../selenium/yohan/data/ham/local.ham1.csv")
df_ham2 = pd.read_csv("../../selenium/yohan/data/ham/local.ham2.csv")
df_ham3 = pd.read_csv("../../selenium/yohan/data/ham/local.ham3.csv")
df_ham4 = pd.read_csv("../../selenium/yohan/data/ham/local.ham4.csv")
df_ham5 = pd.read_csv("../../selenium/yohan/data/ham/local.ham5.csv")

df_kyoja1 = pd.read_csv("../../selenium/yohan/data/kyoja/local.kyoja1.csv")
df_kyoja2 = pd.read_csv("../../selenium/yohan/data/kyoja/local.kyoja2.csv")
df_kyoja3 = pd.read_csv("../../selenium/yohan/data/kyoja/local.kyoja3.csv")
df_kyoja4 = pd.read_csv("../../selenium/yohan/data/kyoja/local.kyoja4.csv")

df_tomato1 = pd.read_csv("../../selenium/yohan/data/tomato_source/local.tomato1.csv")
df_tomato2 = pd.read_csv("../../selenium/yohan/data/tomato_source/local.tomato2.csv")
df_tomato3 = pd.read_csv("../../selenium/yohan/data/tomato_source/local.tomato3.csv")

df_canola1 = pd.read_csv("../../selenium/youjin/data/canola_oil/ssg.canola_oil1.csv")
df_canola2 = pd.read_csv("../../selenium/youjin/data/canola_oil/ssg.canola_oil2.csv")
df_canola3 = pd.read_csv("../../selenium/youjin/data/canola_oil/ssg.canola_oil3.csv")
df_canola4 = pd.read_csv("../../selenium/youjin/data/canola_oil/ssg.canola_oil4.csv")

df_kimchi1 = pd.read_csv("../../selenium/youjin/data/kimchi/ssg.kimchi1.csv")
df_kimchi2 = pd.read_csv("../../selenium/youjin/data/kimchi/ssg.kimchi2.csv")
df_kimchi3 = pd.read_csv("../../selenium/youjin/data/kimchi/ssg.kimchi3.csv")
df_kimchi4 = pd.read_csv("../../selenium/youjin/data/kimchi/ssg.kimchi4.csv")

df_mung_bean_sprouts1 = pd.read_csv("../../selenium/youjin/data/mung_bean_sprouts/ssg.mung_bean_sprouts1.csv")
df_mung_bean_sprouts2 = pd.read_csv("../../selenium/youjin/data/mung_bean_sprouts/ssg.mung_bean_sprouts2.csv")
df_mung_bean_sprouts3 = pd.read_csv("../../selenium/youjin/data/mung_bean_sprouts/ssg.mung_bean_sprouts3.csv")
df_mung_bean_sprouts4 = pd.read_csv("../../selenium/youjin/data/mung_bean_sprouts/ssg.mung_bean_sprouts4.csv")

df_pancake_powder1 = pd.read_csv("../../selenium/youjin/data/pancake_powder/ssg.pancake_powder1.csv")
df_pancake_powder2 = pd.read_csv("../../selenium/youjin/data/pancake_powder/ssg.pancake_powder2.csv")
df_pancake_powder3 = pd.read_csv("../../selenium/youjin/data/pancake_powder/ssg.pancake_powder3.csv")
df_pancake_powder4 = pd.read_csv("../../selenium/youjin/data/pancake_powder/ssg.pancake_powder4.csv")

df_rice1 = pd.read_csv("../../selenium/youjin/data/rice/ssg.rice1.csv")
df_rice2 = pd.read_csv("../../selenium/youjin/data/rice/ssg.rice2.csv")
df_rice3 = pd.read_csv("../../selenium/youjin/data/rice/ssg.rice3.csv")
df_rice4 = pd.read_csv("../../selenium/youjin/data/rice/ssg.rice4.csv")
df_rice5 = pd.read_csv("../../selenium/youjin/data/rice/ssg.rice5.csv")

df_bread1 = pd.read_csv("../../selenium/youngji/data/local.bread_1.csv")
df_bread2 = pd.read_csv("../../selenium/youngji/data/local.bread_2.csv")
df_bread3 = pd.read_csv("../../selenium/youngji/data/local.bread_3.csv")
df_bread4 = pd.read_csv("../../selenium/youngji/data/local.bread_4.csv")

df_cheese1 = pd.read_csv("../../selenium/youngji/data/local.cheese_1.csv")
df_cheese2 = pd.read_csv("../../selenium/youngji/data/local.cheese_2.csv")
df_cheese3 = pd.read_csv("../../selenium/youngji/data/local.cheese_3.csv")

df_milk1 = pd.read_csv("../../selenium/youngji/data/local.milk_1.csv")
df_milk2 = pd.read_csv("../../selenium/youngji/data/local.milk_2.csv")
df_milk3 = pd.read_csv("../../selenium/youngji/data/local.milk_3.csv")

df_water1 = pd.read_csv("../../selenium/youngji/data/local.water_1.csv")
df_water2 = pd.read_csv("../../selenium/youngji/data/local.water_2.csv")
df_water3 = pd.read_csv("../../selenium/youngji/data/local.water_3.csv")

df_yogurt1 = pd.read_csv("../../selenium/youngji/data/local.yogurt_1.csv")
df_yogurt2 = pd.read_csv("../../selenium/youngji/data/local.yogurt_2.csv")
df_yogurt3 = pd.read_csv("../../selenium/youngji/data/local.yogurt_3.csv")

dfs = [df_bibim1,df_bibim2,df_bibim3,df_bibim4,df_oil1,df_oil2,df_oil3,df_oil4,df_ham1,df_ham2,df_ham3,df_ham4,df_ham5,df_kyoja1,df_kyoja2,df_kyoja3,df_kyoja4,df_tomato1,df_tomato2,df_tomato3,df_canola1,df_canola2,df_canola3,df_canola4,df_kimchi1,df_kimchi2,df_kimchi3,df_kimchi4,df_mung_bean_sprouts1,df_mung_bean_sprouts2,df_mung_bean_sprouts3,df_mung_bean_sprouts4,df_pancake_powder1,df_pancake_powder2,df_pancake_powder3,df_pancake_powder4,df_rice1,df_rice2,df_rice3,df_rice4,df_rice5,df_bread1,df_bread2,df_bread3,df_bread4,df_cheese1,df_cheese2,df_cheese3,df_milk1,df_milk2,df_milk3,df_water1,df_water2,df_water3,df_yogurt1,df_yogurt2,df_yogurt3]

## 데이터 전처리

In [3]:
# object로 되어 있는 날짜 데이터를 datetime으로 변환
for i in dfs:
    i['textDate'] = pd.to_datetime(i['textDate'])

In [4]:
# 통합 데이터 제작
df_total = pd.DataFrame()
df_total = pd.concat(dfs)

In [5]:
df_total

_id     merchTitle  merchPrice  merchOld  \
0    65f7b0ff73c6e006405497f1  팔도비빔면 130g*4개        3330       NaN   
1    65f7b0ff73c6e006405497f2  팔도비빔면 130g*4개        3330       NaN   
2    65f7b0ff73c6e006405497f3  팔도비빔면 130g*4개        3330       NaN   
3    65f7b0ff73c6e006405497f4  팔도비빔면 130g*4개        3330       NaN   
4    65f7b0ff73c6e006405497f5  팔도비빔면 130g*4개        3330       NaN   
..                        ...            ...         ...       ...   
850  65f815251d9dbce4e6c2af21  hy 떠먹는 윌 450g        2980       NaN   
851  65f815261d9dbce4e6c2af22  hy 떠먹는 윌 450g        2980       NaN   
852  65f815261d9dbce4e6c2af23  hy 떠먹는 윌 450g        2980       NaN   
853  65f815261d9dbce4e6c2af24  hy 떠먹는 윌 450g        2980       NaN   
854  65f815261d9dbce4e6c2af25  hy 떠먹는 윌 450g        2980       NaN   

     merchEvent    textName   textDate  textLevel  \
0         False  yj9******* 2023-09-16          5   
1         False  dnj******* 2024-03-11          5   
2         False  bhb******* 2024-03-03          5   
3         False  ehk******* 2024-02-03          5   
4         False  ryu******* 2024-01-14          5   
..          ...         ...        ...        ...   
850       False  min******* 2023-05-16          2   
851       False  kjs******* 2023-05-13          2   
852       False  ide******* 2023-07-06          1   
853       False  mjy******* 2023-04-29          1   
854       False  iso******* 2023-05-12          1   

                                           textContent  
0    좋은 상품 할인 받아서 좋은 가격에 구입했습니다 꼼꼼하게 포장되어서 빠르게 배송 받...  
1    이번에 딸기버전으로 뭔가 새로 나왔는지 기본이 아니라 딸기맛으로 왔네요! 근데 딸기...  
2    시즌한정 딸기버젼으로 받았어요. 음... 전 그냥 원래가 더 나은거 같아요. 개인적...  
3             비빔면은 팔도! 소고기에 비빔면 올려 맛있게 먹을 예정. 음-군침 돈다.  
4                                          자주 구매하고 있어요  
..                                                 ...  
850                        유통기한도 짧고 개봉후 하루이틀 지나면 살짝 시다  
851                            너무 기대했나봐요ㅜ마시는 요구르트보다 별로  
852                                    3일남은 재고 보내줌 ㅡ.ㅡ  
853  윌은 그냥 마세는게 100배낫습니다. 궁금해서 샀는데.먹기가.힘드네요. 윌맛에 걸쭉...  
854                          간단하게 이용하기좋아요 간단하게 이용하기좋아요  

[923982 rows x 9 columns]

In [6]:
# 3점을 제외한 나머지 라벨링 작업 함수 제작
def labeling(text):
    if text in [4,5]:
        text = 1
    elif text in [1,2]:
        text = 0
    else:
        pass
    return text

In [7]:
# 학습을 위한 fearure target 데이터 프레임 작성
labeling_data = pd.DataFrame()
labeling_data = df_total[["textLevel","textContent"]]
labeling_data = labeling_data.reset_index()
labeling_data = labeling_data.drop(columns='index')
labeling_data['textLevel'] = labeling_data['textLevel'].apply(labeling)

In [8]:
labeling_data['textLevel'].value_counts()

textLevel
1    904301
3     14772
0      4909
Name: count, dtype: int64

- 약 5천개의 부정문 (데이터 전처리 전) 획득
- 7:3 비율로 긍정문을 랜덤 선별하여 처리해보자

In [9]:
labeling_data

textLevel                                        textContent
0               1  좋은 상품 할인 받아서 좋은 가격에 구입했습니다 꼼꼼하게 포장되어서 빠르게 배송 받...
1               1  이번에 딸기버전으로 뭔가 새로 나왔는지 기본이 아니라 딸기맛으로 왔네요! 근데 딸기...
2               1  시즌한정 딸기버젼으로 받았어요. 음... 전 그냥 원래가 더 나은거 같아요. 개인적...
3               1           비빔면은 팔도! 소고기에 비빔면 올려 맛있게 먹을 예정. 음-군침 돈다.
4               1                                        자주 구매하고 있어요
...           ...                                                ...
923977          0                        유통기한도 짧고 개봉후 하루이틀 지나면 살짝 시다
923978          0                            너무 기대했나봐요ㅜ마시는 요구르트보다 별로
923979          0                                    3일남은 재고 보내줌 ㅡ.ㅡ
923980          0  윌은 그냥 마세는게 100배낫습니다. 궁금해서 샀는데.먹기가.힘드네요. 윌맛에 걸쭉...
923981          0                          간단하게 이용하기좋아요 간단하게 이용하기좋아요

[923982 rows x 2 columns]

In [10]:
# 중복되어 있을 수 있는 내용을 제거
labeling_data.drop_duplicates()['textLevel'].value_counts()

textLevel
1    446016
3      8878
0      3149
Name: count, dtype: int64

- 중복 제거 작업 후 예상보다 상당히 많은 부분이 중복되어 있었음을 확인
    + 긍정문의 약 50%가 중복으로 사라진 것을 보아 중복되고 무의미한 긍정 댓글이 매우 많았음을 확인
    + 이 경우 댓글 알바나 이벤트 등으로 인해 허수였을 것으로 예상할 수 있음 (추후 분석 결과에 추가)

In [11]:
labeling_data = labeling_data.drop_duplicates()

In [12]:
# Null행 확인
labeling_data.isnull().sum()

textLevel      0
textContent    2
dtype: int64

In [13]:
# Null행 제거
labeling_data = labeling_data.dropna()

In [14]:
import re

# 한글과 공백을 제외하고 모두 제거
labeling_data['textContent'] = labeling_data['textContent'].apply(lambda x: re.sub(r"[^ㄱ-ㅎㅏ-ㅣ가-힣 ]", "", x))

In [15]:
import numpy as np

# 스페이스로 이루어진 행을 Null값으로 변경
labeling_data['textContent'] = labeling_data['textContent'].apply(lambda x: re.sub(r'^ +', "", x)) 
labeling_data['textContent'].replace('', np.nan, inplace=True)

C:\Users\서울디지털인재개발원\AppData\Local\Temp\ipykernel_18928\1841035627.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  labeling_data['textContent'].replace('', np.nan, inplace=True)


In [16]:
labeling_data.isnull().sum()

textLevel         0
textContent    1479
dtype: int64

In [17]:
# Null행 제거
labeling_data = labeling_data.dropna()

## Tokenizing

In [18]:
import pickle
import matplotlib.pyplot as plt
import urllib.request
from mecab import MeCab
from tqdm import tqdm
from sklearn.model_selection import train_test_split
from collections import Counter

In [19]:
mecab = MeCab()

In [20]:
# 기본적인 불용어와 상표명 등의 독립단어 설정
stopword1 = ['도', '는', '다', '의', '가', '이', '은', '한', '에', '하', '고', '을', '를', '인', '듯', '과', '와', '네', '들', '듯', '지', '임', '게','의','가','들','는','좀','잘','걍','과','도','를','으로','자','에','와','한','하다']
stopword2 = ['팔도','팔도비빔면','비빔면','괄도','네넴띤','괄도네넴띤','농심','배홍동','배홍동비빔면','오뚜기','진비빔면','삼양','찰비빔면','비비고','교자','왕교자','노브랜드','교자만두','만두','왕교자만두','피코크','담백한왕교자','토마토','청정원','파스타소스','폰타나','샘표','뽀모도로','CJ','cj','스팸','클래식','덴마크튤립햄','튤립햄','동원','리챔','하림','닭가슴살햄','챔','더본코리아','빽햄','백종원','고소한참기름','참기름','경천','경천식품','시골참기름','들기름','최순희','백설','통참깨참기름','통참깨','임금님','임금님표','이천쌀','한눈에 반한 쌀','경기미','농협','농협양곡','당진해나루쌀','삼광','의성','의성농협','바른고을','의성진쌀','김치','별미총각김치','총각김치','종가집','종가','오래오래맛있는총각김치','조선호텔특제육수','조선호텔','부침가루','부침','곰표','통감자','자연주의','숙주나물','친환경','국산숙주','숙주','무농약','통통숙주','카놀라유','카놀라오일','카놀라','Coosur','coosur','쿡투게더','부드러운우유식빵','식빵','우유식빵','삼립','롯데','롯데웰푸드','기린','우유','더 메나쥬리','메나쥬리','온가족우유식빵','삼다수','에비앙','백산수','플레인','요거트','덴마크','에치와이','hy','윌','상하','상하치즈','체다','치즈','서울우유','서울','슬라이스','남양','드빈치','체다치즈','갓밀크','굿모닝','굿밀크','헬로우','앙팡']
stopword = stopword1 + stopword2

In [21]:
train_data = labeling_data.query('textLevel	!= 3')

In [22]:
train_data['textLevel'].value_counts()

textLevel
1    444609
0      3138
Name: count, dtype: int64

In [23]:
feature_train = []
for i in tqdm(train_data['textContent']):
    tokenized = mecab.morphs(i) # mecab을 통한 토큰화
    removed = [j for j in tokenized if not j in stopword] # 불용어 제거
    feature_train.append(removed)

100%|██████████| 447747/447747 [00:49<00:00, 8963.68it/s] 


In [24]:
len(feature_train),len(labeling_data['textLevel'])

(447747, 456562)

## 단어 집합 제작

In [25]:
word_list = []
for i in feature_train:
    for j in i:
      word_list.append(j)

word_counts = Counter(word_list)

In [26]:
threshold = 3
total_cnt = len(word_counts) # 단어의 수
rare_cnt = 0 # 등장 빈도수가 threshold보다 작은 단어의 개수를 카운트
total_freq = 0 # 훈련 데이터의 전체 단어 빈도수 총 합
rare_freq = 0 # 등장 빈도수가 threshold보다 작은 단어의 등장 빈도수의 총 합

# 단어와 빈도수의 쌍(pair)을 key와 value로 받는다.
for key, value in word_counts.items():
    total_freq = total_freq + value

    # 단어의 등장 빈도수가 threshold보다 작으면
    if(value < threshold):
        rare_cnt = rare_cnt + 1
        rare_freq = rare_freq + value

print('단어 집합(vocabulary)의 크기 :',total_cnt)
print('등장 빈도가 %s번 이하인 희귀 단어의 수: %s'%(threshold - 1, rare_cnt))
print("단어 집합에서 희귀 단어의 비율:", (rare_cnt / total_cnt)*100)
print("전체 등장 빈도에서 희귀 단어 등장 빈도 비율:", (rare_freq / total_freq)*100)

단어 집합(vocabulary)의 크기 : 26273
등장 빈도가 2번 이하인 희귀 단어의 수: 15400
단어 집합에서 희귀 단어의 비율: 58.61530849160735
전체 등장 빈도에서 희귀 단어 등장 빈도 비율: 0.5496740253417093


In [27]:
vocab = sorted(word_counts, key=word_counts.get, reverse=True)

In [28]:
# 전체 단어 개수 중 빈도수 2이하인 단어는 제거.
vocab_size = total_cnt - rare_cnt
vocab = vocab[:vocab_size]
print('단어 집합의 크기 :', len(vocab))

단어 집합의 크기 : 10873


In [29]:
word_to_index = {}
word_to_index['<PAD>'] = 0
word_to_index['<UNK>'] = 1

In [30]:
for index, word in enumerate(vocab) :
  word_to_index[word] = index + 2

In [64]:
index_to_word = {}
for key, value in word_to_index.items():
    index_to_word[value] = key

## Vectorization
- TF-IDF 사용

In [31]:
from sklearn.feature_extraction.text import TfidfVectorizer
# tfidfVectorizer = TfidfVectorizer(tokenizer=None,stop_words=stopword, ngram_range=(1,2))
# 모델의 학습의 다향성을 위해 ngram 기법 사용 -> 메모리 문제 발생으로 취소
tfidfVectorizer = TfidfVectorizer(tokenizer=None,stop_words=stopword)

In [32]:
fit_list = []
for i in range(len(feature_train)):
    into_text =''
    for j in range(len(feature_train[i])):
        if word_to_index.get(feature_train[i][j],False):
            into_text += feature_train[i][j]
            if j != len(feature_train[i])-1:
                into_text += ' '
    fit_list.append(into_text)
fit_list

['좋 상품 할인 받 아서 좋 가격 구입 했 습니다 꼼꼼 포장 되 어서 빠르 배송 받 았 습니다 유통 기한 넉넉 해서 좋 아요',
 '이번 딸기 버전 뭔가 새로 나왔 는지 기본 아니 라 딸기 맛 왔 네요 근데 딸기 맛 후 첨 분말 빼 면 그냥 원래 랑 똑같 아서 빼 먹 었 습니다 딸기 맛 도전 해 보 기 엔 음 쓰 서 ㅎㅎ',
 '시즌 한정 딸기 버 젼 받 았 어요 음 전 그냥 원래 더 나 거 같 아요 개인 적 그 딸기 향 단맛 살짝 추가 된 맛 전 단맛 보다 그 새콤 맛 좋 아서요',
 '팔 소고기 올려 맛있 먹 예정 음 군침 돈다',
 '자주 구매 있 어요',
 '쟁여 놓 품목 중 하나 입니다 맛있 는데 가격 올라서 세일 이용 해서 주문 합니다 소스 달콤 적당히 맵 콤 해서 반찬 없 때 밥맛 없 때 더울 때 땡길 때 자주 해 먹 어요 추천 해요',
 '뭐 명불허전 죠 근데 쓱 배송 시켰 는대 도착 클라스 크으 으',
 '우리 집 없 으면 안 되 역시 팔 지점 마다 다를 수 잇 겟 지만 여긴 왜 맨날 거의 뿌셔뿌셔 마냥 이렇게 오 는지 정말',
 '에서 먹 거 보 바로 주문 오 배송 되 어서 딸기 맛 ㅜㅜ 더 비싼데 ㅠㅠ 딸기 맛 왔 능',
 '제 입맛 엔 딱 거 같 아요 그런데 얼마 전 마요네즈 섞 어 먹 었 는데 너무 맛있 거 에요 매운 걸 못 먹 편 인데 훨씬 부드럽 중독 성 있 맛 되 더라구요 추천 합니다',
 '최고 입니다 쿠 팡 쓰 질려서 이젠 이마트 홈 플 만 씁니다 상품 다양 해서 좋 새벽 주문 당일 도착 최고 요',
 '늘 주문 거 에요 성비 좋 맛있 어서 계속 구입 됩니다 앞 재고 항상 있 었 으면 좋 겠 어요',
 '정석 팔 입니다 배송 빨라서 좋 았 맛 말 할 필요 없 습니다 겨울 지만 갑자기 땡겨 서 구매 했 어요 먹 겠 습니다',
 '자세히 안 봤 는데 개 아니 라 개입 네요 겨울 용 육수 스프 추가 로 어 있 어 국물 같이 먹 으니 좋 네요',
 '윈터 에디션 와서 너무 기분 좋 았 습니다 ㅎㅎ 항상 시켜 먹 는데 뭐니뭐니 해도 팔

In [33]:
# 잠깐! 이거 좀 더 생각해보고 하자.
# 타겟을어떻게 해야되는지 아직 정리가 안됐음
get_vector = tfidfVectorizer.fit_transform(fit_list)


C:\Users\서울디지털인재개발원\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\feature_extraction\text.py:408: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['반한', '한눈에'] not in stop_words.
  warnings.warn(


In [34]:
get_vector.shape,train_data['textLevel'].shape

((447747, 9510), (447747,))

## 학습 데이터 / 검증 데이터 / 테스트 데이터 분리

### 데이터 비율 맞추기
    - 현재 데이터의 비율이 매우 차이가 심함
    - 비정형 데이터(자연어 처리)이기 때문에 가중치 조절이나 Oversampling은 하지 않는 것이 바람직
    - UnderSampling진행

In [35]:
from imblearn.under_sampling import NearMiss
underSampling = NearMiss(sampling_strategy=0.7)
feature_undersample, target_undersample = underSampling.fit_resample(get_vector,train_data['textLevel'])
feature_undersample.shape, target_undersample.shape

((7620, 9510), (7620,))

In [36]:
target_undersample.value_counts()

textLevel
1    4482
0    3138
Name: count, dtype: int64

- 긍정문 약 4천개
- 부정문 약 3천개

### 데이터 스플릿

In [37]:
feature_train, feature_test, target_train, target_test = train_test_split(feature_undersample,target_undersample, test_size=0.2, random_state=42, stratify=target_undersample)


In [38]:
feature_train.shape,target_train.shape

((6096, 9510), (6096,))

In [39]:
feature_train, feature_valid, target_train, target_valid = train_test_split(feature_train,target_train,test_size=0.2, random_state=42, stratify=target_train)
feature_train.shape,feature_valid.shape,feature_test.shape,target_train.shape,target_valid.shape,target_test.shape

((4876, 9510), (1220, 9510), (1524, 9510), (4876,), (1220,), (1524,))

- 학습 데이터 약 5천개
- 검증 데이터 약 천개
- 테스트 데이터 약 천5백개

In [40]:
print('--------학습 데이터의 비율-----------')
print(f'부정 리뷰 = {round(np.sum(target_train==0)/len(target_train) * 100,3)}%')
print(f'긍정 리뷰 = {round(np.count_nonzero(target_train)/len(target_train) * 100,3)}%')
print('--------검증 데이터의 비율-----------')
print(f'부정 리뷰 = {round(np.sum(target_valid==0)/len(target_valid) * 100,3)}%')
print(f'긍정 리뷰 = {round(np.count_nonzero(target_valid)/len(target_valid) * 100,3)}%')
print('--------테스트 데이터의 비율-----------')
print(f'부정 리뷰 = {round(np.sum(target_test==0)/len(target_test) * 100,3)}%')
print(f'긍정 리뷰 = {round(np.count_nonzero(target_test)/len(target_test) * 100,3)}%')

--------학습 데이터의 비율-----------
부정 리뷰 = 41.181%
긍정 리뷰 = 58.819%
--------검증 데이터의 비율-----------
부정 리뷰 = 41.148%
긍정 리뷰 = 58.852%
--------테스트 데이터의 비율-----------
부정 리뷰 = 41.207%
긍정 리뷰 = 58.793%


## 모델 학습
- LSTM

In [41]:
import torch
import torch.nn as nn
import torch.nn.functional as F

In [42]:
USE_CUDA = torch.cuda.is_available()
device = torch.device("cuda" if USE_CUDA else "cpu")
print("cpu와 cuda 중 다음 기기로 학습함:", device)

cpu와 cuda 중 다음 기기로 학습함: cpu


In [43]:
train_label_tensor = torch.tensor(np.array(target_train))
valid_label_tensor = torch.tensor(np.array(target_valid))
test_label_tensor = torch.tensor(np.array(target_test))
print(train_label_tensor[:5])

tensor([1, 1, 1, 0, 0])


In [52]:
class TextClassifier(nn.Module):
    def __init__(self, vocab_size, embedding_dim, hidden_dim, output_dim):
        super(TextClassifier, self).__init__()
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, batch_first=True)
        
        # 메모리 문제로 모델 축소....
        # self.lstm = nn.LSTM(embedding_dim, hidden_dim*2, batch_first=True)
        # self.lstm2 = nn.LSTM(hidden_dim*2, hidden_dim, batch_first=True)

        self.fc = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        # x: (batch_size, seq_length)
        embedded = self.embedding(x)  # (batch_size, seq_length, embedding_dim)

        # LSTM은 (hidden state, cell state)의 튜플을 반환합니다
        lstm_out, (hidden, cell) = self.lstm(embedded)  # lstm_out: (batch_size, seq_length, hidden_dim), hidden: (1, batch_size, hidden_dim)

        '''
        # 메모리 문제....
        # lstm_out2, _ = self.lstm2(lstm_out1)  # lstm_out2: (batch_size, seq_length, hidden_dim)
        # last_hidden = lstm_out2[:, -1, :]  # 마지막 시퀀스에서의 hidden state만 추출합니다
        '''
        last_hidden = hidden.squeeze(0)  # (batch_size, hidden_dim)
        logits = self.fc(last_hidden)  # (batch_size, output_dim)
        # softmax 함수를 적용하여 확률로 변환
        probs = F.softmax(logits, dim=1)
        return probs

In [53]:
encoded_train = torch.tensor(feature_train.toarray()).to(torch.int64)
train_dataset = torch.utils.data.TensorDataset(encoded_train, train_label_tensor)
train_dataloader = torch.utils.data.DataLoader(train_dataset, shuffle=True, batch_size=32)

encoded_test = torch.tensor(feature_test.toarray()).to(torch.int64)
test_dataset = torch.utils.data.TensorDataset(encoded_test, test_label_tensor)
test_dataloader = torch.utils.data.DataLoader(test_dataset, shuffle=True, batch_size=1)

encoded_valid = torch.tensor(feature_valid.toarray()).to(torch.int64)
valid_dataset = torch.utils.data.TensorDataset(encoded_valid, valid_label_tensor)
valid_dataloader = torch.utils.data.DataLoader(valid_dataset, shuffle=True, batch_size=1)

In [54]:
total_batch = len(train_dataloader)
print('총 배치의 수 : {}'.format(total_batch))

총 배치의 수 : 153


In [55]:
embedding_dim = 100
hidden_dim = 128
output_dim = 2
learning_rate = 0.001
num_epochs = 10

model = TextClassifier(vocab_size, embedding_dim, hidden_dim, output_dim)
model.to(device)

TextClassifier(
  (embedding): Embedding(10873, 100)
  (lstm): LSTM(100, 128, batch_first=True)
  (fc): Linear(in_features=128, out_features=2, bias=True)
)

In [56]:
# 크로스엔트로피 손실함수 사용
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

### 평가  코드

In [57]:
def calculate_accuracy(logits, labels):
    predicted = torch.argmax(logits, dim=1)
    correct = (predicted == labels).sum().item()
    total = labels.size(0)
    accuracy = correct / total
    return accuracy

In [58]:
def evaluate(model, valid_dataloader, criterion, device):
    val_loss = 0
    val_correct = 0
    val_total = 0

    model.eval()
    with torch.no_grad():
        # 데이터로더로부터 배치 크기만큼의 데이터를 연속으로 로드
        for batch_X, batch_y in valid_dataloader:
            batch_X, batch_y = batch_X.to(device), batch_y.to(device)

            # 모델의 예측값
            logits = model(batch_X)

            # 손실을 계산
            loss = criterion(logits, batch_y)

            # 정확도와 손실을 계산함
            val_loss += loss.item()
            val_correct += calculate_accuracy(logits, batch_y) * batch_y.size(0)
            val_total += batch_y.size(0)

    val_accuracy = val_correct / val_total
    val_loss /= len(valid_dataloader)

    return val_loss, val_accuracy

## 학습 진행

In [60]:
# Training loop
best_val_loss = float('inf')

# Training loop
for epoch in range(num_epochs):
    # Training
    train_loss = 0
    train_correct = 0
    train_total = 0
    model.train()
    for batch_X, batch_y in train_dataloader:
        # Forward pass
        batch_X, batch_y = batch_X.to(device), batch_y.to(device)
        # batch_X.shape == (batch_size, max_len)
        logits = model(batch_X)

        # Compute loss
        loss = criterion(logits, batch_y)

        # Backward pass and optimization
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        # Calculate training accuracy and loss
        train_loss += loss.item()
        train_correct += calculate_accuracy(logits, batch_y) * batch_y.size(0)
        train_total += batch_y.size(0)

    train_accuracy = train_correct / train_total
    train_loss /= len(train_dataloader)

    # Validation
    val_loss, val_accuracy = evaluate(model, valid_dataloader, criterion, device)

    print(f'Epoch {epoch+1}/{num_epochs}:')
    print(f'Train Loss: {train_loss:.4f}, Train Accuracy: {train_accuracy:.4f}')
    print(f'Validation Loss: {val_loss:.4f}, Validation Accuracy: {val_accuracy:.4f}')

    # 검증 손실이 최소일 때 체크포인트 저장
    if val_loss < best_val_loss:
        print(f'Validation loss improved from {best_val_loss:.4f} to {val_loss:.4f}. 체크포인트를 저장합니다.')
        best_val_loss = val_loss
        torch.save(model.state_dict(), 'best_model_checkpoint.pth')

Epoch 1/10:
Train Loss: 0.6785, Train Accuracy: 0.5882
Validation Loss: 0.6797, Validation Accuracy: 0.5885
Validation loss improved from inf to 0.6797. 체크포인트를 저장합니다.
Epoch 2/10:
Train Loss: 0.6783, Train Accuracy: 0.5882
Validation Loss: 0.6778, Validation Accuracy: 0.5885
Validation loss improved from 0.6797 to 0.6778. 체크포인트를 저장합니다.
Epoch 3/10:
Train Loss: 0.6772, Train Accuracy: 0.5882
Validation Loss: 0.6775, Validation Accuracy: 0.5885
Validation loss improved from 0.6778 to 0.6775. 체크포인트를 저장합니다.
Epoch 4/10:
Train Loss: 0.6786, Train Accuracy: 0.5882
Validation Loss: 0.6776, Validation Accuracy: 0.5885
Epoch 5/10:
Train Loss: 0.6780, Train Accuracy: 0.5882
Validation Loss: 0.6775, Validation Accuracy: 0.5885
Validation loss improved from 0.6775 to 0.6775. 체크포인트를 저장합니다.
Epoch 6/10:
Train Loss: 0.6776, Train Accuracy: 0.5882
Validation Loss: 0.6777, Validation Accuracy: 0.5885
Epoch 7/10:
Train Loss: 0.6778, Train Accuracy: 0.5882
Validation Loss: 0.6775, Validation Accuracy: 0.5885

## 모델 로드 및 평가

In [61]:
# 모델 로드
model.load_state_dict(torch.load('best_model_checkpoint.pth'))

# 모델을 device에 올립니다.
model.to(device)

TextClassifier(
  (embedding): Embedding(10873, 100)
  (lstm): LSTM(100, 128, batch_first=True)
  (fc): Linear(in_features=128, out_features=2, bias=True)
)

In [62]:
# 검증 데이터에 대한 정확도와 손실 계산
val_loss, val_accuracy = evaluate(model, valid_dataloader, criterion, device)

print(f'Best model validation loss: {val_loss:.4f}')
print(f'Best model validation accuracy: {val_accuracy:.4f}')

Best model validation loss: 0.6774
Best model validation accuracy: 0.5885


In [63]:
# 테스트 데이터에 대한 정확도와 손실 계산
test_loss, test_accuracy = evaluate(model, test_dataloader, criterion, device)

print(f'Best model test loss: {test_loss:.4f}')
print(f'Best model test accuracy: {test_accuracy:.4f}')

Best model test loss: 0.6776
Best model test accuracy: 0.5879


## 모델 테스트

In [96]:
index_to_tag = {0 : '부정', 1 : '긍정'}

def predict(text, model, word_to_index, index_to_tag):
    # Set the model to evaluation mode
    model.eval()

    # Tokenize the input text
    tokens = mecab.morphs(text) # 토큰화
    tokens = [word for word in tokens if not word in stopword] # 불용어 제거
    token_indices = [word_to_index.get(token, 1) for token in tokens]

    # Convert tokens to tensor
    input_tensor = torch.tensor([token_indices], dtype=torch.long).to(device)  # (1, seq_length)

    # Pass the input tensor through the model
    with torch.no_grad():
        logits = model(input_tensor)  # (1, output_dim)

    # Get the predicted class index
    predicted_index = torch.argmax(logits, dim=1)

    # Convert the predicted index to its corresponding tag
    predicted_tag = index_to_tag[predicted_index.item()]

    return predicted_tag

In [101]:
test_input = "맛있어요"
predict(test_input, model, word_to_index, index_to_tag)

'긍정'

In [105]:
test_input = "맛없음"
predict(test_input, model, word_to_index, index_to_tag)

'긍정'

## 결론

- 검증과 테스트의 결과 약 68%의 성능을 보임
- 성능의 하락의 원인으로는 데이터의 부족이 제일 크다고 생각됨
    + 초기 긍,부정의 비율이 매우 치우처져있어서 충분한 학습을 하지 못한 것으로 판단
    + 이를 보완하기 위해 다방향의 부정문 데이터를 더욱 수집할 필요가 있을 것
- 적은량의 데이터로 최대한의 효율을 뽑기 위해 모델의 복잡도와 ngram을 진행했으나 컴퓨터 성능의 한계로 시험해보지 못함
    + 여러번 테스트와 모델의 복잡성을 통해 제작 해야할 것으로 생각됨